### PR TITLE
feat: first draft of JMS Binding Objects

### PR DESCRIPTION
**Pull request by adamretter**
Resolves https://github.com/asyncapi/spec/issues/919

This is a first draft of a set of JMS Binding Objects. There is a lot of variance between JMS Providers in terms of features and capabilities, however these bindings should work with all providers as they stay within the JMS API.
More features could of course be added in future if desirable!

### CODE REVIEW

**Comment by rwalpole** [comment]
It would be great to see this PR merged. We are using AsyncAPI together with JMS in a major project right now.

File: jms/README.md
Comment (derberg) [inline jms/README.md:24]: do we need this section? `Server Object` is not part of the binding. I think just remove it and only provide full example of `Server Object` with binding in `Server Binding Object` section

File: jms/README.md
Comment (derberg) [inline jms/README.md:98]: can you provide some example of a message, the best would be with headers both for the app (`message.headers`) and the protocol headers

**Review by derberg — CHANGES_REQUESTED** [review]
Thanks a lot! Left just 2 comments for now

File: jms/README.md
Comment (adamretter) [reply jms/README.md:98]: @derberg Sure! I just appended a commit which provides such an example :-)

File: jms/README.md
Comment (adamretter) [reply jms/README.md:46]: Sure! I just appended a commit which provides such an example :-)

File: jms/README.md
Comment (adamretter) [reply jms/README.md:82]: I simply followed the AnypointMQ binding as a template for this - https://github.com/asyncapi/bindings/tree/master/anypointmq#examples

I will append a commit to remove the second channel in the example...

Comment (adamretter) [reply jms/README.md:82]: @KhudaDad414 I have appended a commit to remove the second channel from the example.

File: jms/json_schemas/message.json
Comment (adamretter) [reply jms/json_schemas/message.json:27]: I think there is value in this, as the example is also a much simpler reference for the de-facto JMS Headers that are available according to the JMS API Spec - https://docs.oracle.com/javaee/7/api/javax/jms/Message.html.
This list of possible headers does not appear anywhere else in the AsyncAPI binding.

Does that make sense? If not, let me know and I will shorten the example.

File: jms/README.md
Comment (adamretter) [reply jms/README.md]: Oops. Thanks I have appended a commit to fix that - it should be an array.

**Comment by adamretter** [comment]
@derberg @KhudaDad414 @rwalpole I hope I have addressed all of the review comments that you had. Thanks for the feedback :-) I think it is now ready for a second round of review...

**Comment by derberg** [comment]
LGTM but from the perspective of bindings repo maintainer, not `jms` expert. I know zero about the subject

@KhudaDad414 please have another look on JSON Schemas

@rwalpole anything else from your side

@MichaelDavisSolace I think you use JMS at Solace afaik...at least I think you asked about JMS binding one time. Would you have time to look at this PR?

@rcarrascosps you created https://github.com/asyncapi/bindings/issues/87 so maybe you want to join review?

@fmvilas anything from your side?

---
@adamretter Let's give these folks a week for reply. In the meantime can you clarify if you can become a maintainer of this specific binding? just like we have others in https://github.com/asyncapi/bindings/blob/master/CODEOWNERS

**Review by rwalpole — APPROVED** [review]

**Review by KhudaDad414 — APPROVED** [review]
Schemas seems to follow the documentation. 👍

**Comment by adamretter** [comment]
> In the meantime can you clarify if you can become a maintainer of this specific binding?

@derberg This work was done on behalf of one of my clients; so I can maintain it only for as long as they are asking me to do so.

**Review by rcoppen — APPROVED** [review]

**Review by rcoppen — DISMISSED** [review]

**Comment by adamretter** [comment]
> Let's give these folks a week for reply

@derberg It's now been two weeks and I have also addressed some suggestions from @rcoppen.
Do you think this might now be ready to merge please?

**Review by KhudaDad414 — DISMISSED** [review]

**Review by dalelane — APPROVED** [review]

**Review by derberg — APPROVED** [review]
Amazing work @adamretter 🚀 thanks a bunch 🙏🏼 

PR is here for over a month, last commit a week ago and no more requests for change.

The most important question is, are you willing to maintain it?
Basically be a gatekeeper for followup changes?

In this repo some bindings have defined maintainers -> https://github.com/asyncapi/bindings/blob/master/CODEOWNERS

Can you and do you want to become one?

**Comment by adamretter** [comment]
Thanks @derberg :-)
I did previously already answer your questions about becoming a maintainer here - https://github.com/asyncapi/bindings/pull/193#issuecomment-1526060999

**Comment by derberg** [comment]
oh sorry, missed that

so please update this PR with modification in https://github.com/asyncapi/bindings/blob/master/CODEOWNERS

add:
```
/jms/ @adamretter
```

and remember to notify us if you no longer can respond to PRs and issues, best do it by opening an issue that you have to step down and new volunteers are needed. This is very important to remember long term 🙏🏼

**Comment by derberg** [comment]
> I am being pressured into a position that I do not feel comfortable agreeing to support.

Hey, no pressure 😆 I'm not native speaker, when I read `so I can maintain it only for as long as they are asking me to do so` I understand that temporarily you can be a maintainer 🤷🏼 

I'm long enough in open source maintenance to be happy not only about features contributed, but also when I know who will be able to respond to issues and PRs a day after the merge 😄 cause in case of JMS, I'm definitely a last resort 🙃  

Anyway, I definitely recommend you educate your client on open source sustainability and that maintenance support is equally important to feature development support.

---

@rcoppen @dalelane @KhudaDad414 @rwalpole you all were involved in review of the PR. Any of you think you can help with maintenance of the binding. Some of you maintain other bindings already. If we have at least 2 we are in a perfect situation

**Comment by SrfHead** [comment]
Hi all, I have spoken to @rcoppen about this today and I would be happy to take on the maintenance of the bindings. I am currently working on the JMS client within MQ so this is an area I'm interested in helping establish.

**Review by rwalpole — APPROVED** [review]

**Comment by derberg** [comment]
@SrfHead awesome, thanks!

@rcoppen as experienced bindings maintainer would you be willing to join @SrfHead that is new to AsyncAPI community. You both would be best combination as owners of the binding

**Comment by adamretter** [comment]
Any updates on this please?

**Comment by adamretter** [comment]
Any updates on this please?

**Comment by adamretter** [comment]
@derberg any updates on this please?

**Comment by derberg** [comment]
@adamretter sorry, was out on holidays. I see that @rcoppen have a 👍🏼 and @SrfHead is ready to be codeowner here, so please go ahead and update in the PR the `CODEOWNERS` file to add them as owners of JMS binding.

**Comment by adamretter** [comment]
@derberg Thanks. I hope you had a nice holiday? I have now committed the `CODEOWNERS` as you requested!

**Review by derberg — APPROVED** [review]

**Comment by derberg** [comment]
@KhudaDad414 please approve again. No changes to the binding, @adamretter was just adding `CODEOWNERS`

**Comment by derberg** [comment]
@jonaslagoni we are letting in a new binding. Anything we should remember to do because of coming 3.0?

get schemas in `spec-json-schemas` repo, and what else?

**Review by KhudaDad414 — APPROVED** [review]

**Comment by jonaslagoni** [comment]
> @jonaslagoni we are letting in a new binding. Anything we should remember to do because of coming 3.0?
>
> get schemas in spec-json-schemas repo, and what else?

Nothing tbh, you just have the same question as I gave the other codeowners @adamretter @rcoppen: https://github.com/asyncapi/bindings/issues/182#issuecomment-1589209363

You can find a lot of information about the changes here: https://github.com/asyncapi/bindings/issues/182#issuecomment-1594663850

Also happy to help out, and feel free to join our spec 3.0 meetings if you have questions you want answered live 🙂

**Comment by derberg** [comment]
@adamretter @rcoppen @SrfHead as per @jonaslagoni comment. In short, this binding that was just merged is for AsyncAPI 2.x and you need to make sure it will still be working fine with 3.x. So look on linked materials and any changes need to be done against `next-major-spec` branch.

also, in 3.0 we remove json schemas from this repo, to have them better integrated with current tooling, we move them to https://github.com/asyncapi/spec-json-schemas. So please open up a followup PR to get your bindings in https://github.com/asyncapi/spec-json-schemas/tree/next-major-spec/bindings + make sure they are referenced in respective `*BindingsObject.json` files in https://github.com/asyncapi/spec-json-schemas/tree/next-major-spec/definitions/3.0.0

**Pull request merged**
